thentos-adhocracy/src/Thentos/Adhocracy3/Types.hs

```diff
@@ -82,20 +85,81 @@ instance SafeCopy (ThentosError DB)
+    toJSON err = object
+        [ "status" .= String "error"
+        , "errors" .= [ object
+            [ "name"        .= aeName err
+            , "location"    .= aeLocation err
+            , "description" .= aeDescription err
+            ]]
+        ]
+
+-- Construct a simple A3-style error, with 'aeName' set to "thentos" and 'aeLocation' set to "body".
+-- Useful for cases where all we really have is a description.
+mkSimpleA3ErrorInfo :: ST -> A3ErrorInfo
+mkSimpleA3ErrorInfo desc = A3ErrorInfo
+    {aeName = "thentos", aeLocation = "body", aeDescription = desc}
+
+-- | Construct a ServantErr that looks like those reponrted by the A3 backend.
+-- The backend returns a list of errors but we always use a single-element list, as Thentos
+-- aborts at the first detected error.
+mkA3StyleServantErr :: ServantErr -> A3ErrorInfo -> ServantErr
+mkA3StyleServantErr baseErr errorInfo = baseErr
+    {errBody = encode errorInfo, errHeaders = [contentTypeJsonHeader]}
+
+-- | Construct a simple A3-style 'ServantErr' from only a base error and a message, setting the
+-- other A3-specific fields to default values.
+mkSimpleA3StyleServantErr :: MkServantErrFun
+mkSimpleA3StyleServantErr baseErr msg = mkA3StyleServantErr baseErr $ mkSimpleA3ErrorInfo msg
+
 instance ThentosErrorToServantErr DB where
-    thentosErrorToServantErr (ThentosA3ErrorCore e) = thentosErrorToServantErr e
-    thentosErrorToServantErr e@(A3BackendErrorResponse _ _) =
-        (Just (ERROR, show e), mkServantErr err500 "exception in a3 backend")
-
-    thentosErrorToServantErr e@(A3BackendInvalidJson _) = do
-        (Just (ERROR, show e), mkServantErr err500 "exception in a3 backend: received bad json")
-    -- the following shouldn't actually reach servant:
-    thentosErrorToServantErr e@SsoErrorUnknownCsrfToken =
-        (Just (ERROR, show e), mkServantErr err500 "invalid token returned during sso process")
-    thentosErrorToServantErr e@(SsoErrorCouldNotAccessUserInfo _) =
-        (Just (ERROR, show e), mkServantErr err500 "error accessing user info")
-    thentosErrorToServantErr e@(SsoErrorCouldNotGetAccessToken _) =
-        (Just (ERROR, show e), mkServantErr err500 "error retrieving access token")
+    thentosErrorToServantErr mMkErr = f
+      where
+        mkErr = fromMaybe mkSimpleA3StyleServantErr mMkErr
+
+        -- For errors specifically relevant to the A3 frontend we mirror the A3 backend errors
+        -- exactly so that the frontend recognizes them
+        f (ThentosA3ErrorCore e) = case e of
+            BadCredentials -> (Nothing, mkA3StyleServantErr err400 $ A3ErrorInfo
+                "password"
+                "body"
+                "User doesn't exist or password is wrong")
+            UserEmailAlreadyExists -> (Nothing, mkA3StyleServantErr err400 $ A3ErrorInfo
+                "data.adhocracy_core.sheets.principal.IUserExtended.email"
+                "body"
+                "The user login email is not unique")
```

Comment:
I test the conversion of `BadCredentials` and `NoSuchPendingUserConfirmation`, not of all the A3-specific error types. If those two work, it's reasonable to assume that the others will work as well.